File: android-player/app/src/main/java/jp/virtualcd/player/case3d/DiscPullGesture.java
package jp.virtualcd.player.case3d;

/** Two-finger hub/edge grab. Capture lasts until all fingers lift, even if cancelled. */
public final class DiscPullGesture {
    private int hub=-1,edge=-1;
    private float hx,hy,ex,ey;
    private boolean captured,finished;
    public boolean begin(int a,float ax,float ay,float ar,int b,float bx,float by,float br){
        reset();
        if(ar<=.28f&&ar>=0&&br>=.70f&&br<=1.15f){hub=a;edge=b;hx=ax;hy=ay;ex=bx;ey=by;}
        else if(br<=.28f&&br>=0&&ar>=.70f&&ar<=1.15f){hub=b;edge=a;hx=bx;hy=by;ex=ax;ey=ay;}
        else return false;
        captured=true;return true;
    }
    public boolean move(int a,float ax,float ay,int b,float bx,float by,float density){
        if(!captured||finished)return false;
        if(a==edge&&b==hub)return move(b,bx,by,a,ax,ay,density);
        if(a!=hub||b!=edge){cancel();return false;}
        if(Math.hypot(ax-hx,ay-hy)>20*density){cancel();return false;}
        if(Math.hypot(bx-ex,by-ey)<28*density)return false;
        finished=true;return true;
    }
    public boolean captured(){return captured;}
    public void cancel(){finished=true;}
    public void reset(){captured=finished=false;hub=edge=-1;}
}

File: DiscPullGesture.cs
using System.Windows;
namespace ZipMp3Player;

internal sealed class DiscPullGesture
{
    private int hub,edge;
    private Point hubStart,edgeStart;
    private bool finished;
    public bool Captured{get;private set;}
    public bool Begin(int a,Point ap,double ar,int b,Point bp,double br){
        Reset();
        if(ar>=0&&ar<=.28&&br>=.70&&br<=1.15){hub=a;edge=b;hubStart=ap;edgeStart=bp;}
        else if(br>=0&&br<=.28&&ar>=.70&&ar<=1.15){hub=b;edge=a;hubStart=bp;edgeStart=ap;}
        else return false;
        return Captured=true;
    }
    public bool Move(int a,Point ap,int b,Point bp){
        if(!Captured||finished)return false;
        if(a==edge&&b==hub)return Move(b,bp,a,ap);
        if(a!=hub||b!=edge||(ap-hubStart).Length>20){Cancel();return false;}
        if((bp-edgeStart).Length<28)return false;
        finished=true;return true;
    }
    public void Cancel()=>finished=true;
    public void Reset(){Captured=false;finished=false;}
}
